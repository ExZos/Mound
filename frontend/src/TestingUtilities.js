import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'user' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
