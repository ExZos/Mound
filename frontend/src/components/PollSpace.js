import React from 'react';
import { Button, ButtonGroup, CircularProgress } from '@material-ui/core';

import GeneralComponent from './GeneralComponent';
import { server, api } from '../server';
import '../styles/pollSpace.scss';

class PollSpace extends GeneralComponent {
  constructor(props) {
    super(props);

    this.loaded = false;

    this.state = {
      polls: [],
      vote: {}
    };
  }

  componentDidMount() {
    this.getPolls();
    this.interval = setInterval(this.getPolls, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  getPolls = async () => {
    try {
      const res = await server.get(api.getPendingUnvotedPollsForUser + this.props.userID);

      this.loaded = true;

      this.setState({
        polls: res.data
      });
    } catch (e) {
      // TODO: render error component
    }
  }

  addVote = async (e, result) => {
    const vote = {
      poll: e.target.id,
      user: this.props.userID,
      result: result
    };

    try {
      /*const res = */await server.post(api.createVoteNUpdatePoll, vote);

      this.getPolls();
    } catch (e) {
      // TODO: render error component
    }
  }

  determinePollStatement = (poll) => {
    if(!poll.user) {
      return (
        <div>
          <span className="userName">{poll.name}</span> wants to join this space.
        </div>
      );
    }
    else if(poll.name) {
      return "Name Poll";
    }

    return "Ban Poll";
  }

  renderPolls = () => {
    return this.state.polls.map(poll => (
      <div key={poll.id} className="poll">
        <div className="pollStatement">
          {this.determinePollStatement(poll)}
        </div>

        <ButtonGroup variant="contained" className="pollAnswers">
          <Button className="accept" tabIndex="-1" id={poll.id} onClick={(e) => this.addVote(e, true)}>Accept</Button>
          <Button className="decline" tabIndex="-1" id={poll.id} onClick={(e) => this.addVote(e, false)}>Decline</Button>
        </ButtonGroup>
      </div>
    ));
  }

  render() {
    if(!this.loaded) {
      return (
        <div id="pollSpace">
          <CircularProgress color="inherit" />
        </div>
      );
    }
    else if(this.state.polls.length > 0) {
      return (
        <div id="pollSpace">
          {this.renderPolls()}
        </div>
      );
    }

    return (
      <div id="pollSpace">
        No polls
      </div>
    );
  }
}

export default PollSpace;
