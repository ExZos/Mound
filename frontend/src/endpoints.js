export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
