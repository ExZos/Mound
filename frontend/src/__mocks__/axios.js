import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
