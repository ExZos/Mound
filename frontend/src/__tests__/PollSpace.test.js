import React from 'react';
import { shallow, mount, render } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import PollSpace from '../components/PollSpace';
import TestingUtilities from '../TestingUtilities';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('PollSpace', () => {
  it('should render no polls correctly with props', async () => {
    const component = await shallow(<PollSpace userID={1} />);
    const polls = component.find('div.poll');

    expect(component).toMatchSnapshot();
    expect(polls).toHaveLength(0);
  });

  it('should render poll correctly with props', async () => {
    const component = await shallow(<PollSpace userID={2} />);
    const polls = component.find('div.poll');

    expect(component).toMatchSnapshot();
    expect(polls).toHaveLength(3);
    expect(polls.at(0).text()).toContain('wants to join');
    expect(polls.at(1).text()).toContain('Name Poll');
    expect(polls.at(2).text()).toContain('Ban Poll');
  });

  it('should call post api on accept click with positive vote', async () => {
    const component = await shallow(<PollSpace userID={2} />);
    const button = component.find('Button.accept');
    const spyPost = jest.spyOn(axios, 'post');
    const spyGet = jest.spyOn(axios, 'get');

    button.at(0).props().onClick({
      target: {
        id: 2
      }
    });

    expect(spyPost).toHaveBeenCalled();
    expect(spyPost).toHaveBeenLastCalledWith(api.createVoteNUpdatePoll, {
      poll: 2,
      user: 2,
      result: true
    });
    expect(await spyGet).toHaveBeenCalled();
  });

  it('should call post api on decline click with negative vote', async () => {
    const component = await shallow(<PollSpace userID={2} />);
    const button = component.find('Button.decline');
    const spyPost = jest.spyOn(axios, 'post');
    const spyGet = jest.spyOn(axios, 'get');

    button.at(1).props().onClick({
      target: {
        id: 2
      }
    });

    expect(spyPost).toHaveBeenCalled();
    expect(spyPost).toHaveBeenLastCalledWith(api.createVoteNUpdatePoll, {
      poll: 2,
      user: 2,
      result: false
    });
    expect(await spyGet).toHaveBeenCalled();
  });
});
