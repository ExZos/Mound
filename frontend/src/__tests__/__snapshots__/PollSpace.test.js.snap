// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollSpace should render no polls correctly with props 1`] = `
<div
  id="pollSpace"
/>
`;

exports[`PollSpace should render poll correctly with props 1`] = `
<div
  id="pollSpace"
>
  <div
    className="poll"
    key="1"
  >
    <div
      className="pollStatement"
    >
      <div>
        <span
          className="userName"
        >
          user1
        </span>
         wants to join this space.
      </div>
    </div>
    <div
      className="pollAnswers"
    >
      <Button
        className="accept"
        color="success"
        id={1}
        onClick={[Function]}
        tag="button"
      >
        Accept
      </Button>
      <Button
        className="decline"
        color="danger"
        id={1}
        onClick={[Function]}
        tag="button"
      >
        Decline
      </Button>
    </div>
  </div>
  <div
    className="poll"
    key="2"
  >
    <div
      className="pollStatement"
    >
      Name Poll
    </div>
    <div
      className="pollAnswers"
    >
      <Button
        className="accept"
        color="success"
        id={2}
        onClick={[Function]}
        tag="button"
      >
        Accept
      </Button>
      <Button
        className="decline"
        color="danger"
        id={2}
        onClick={[Function]}
        tag="button"
      >
        Decline
      </Button>
    </div>
  </div>
  <div
    className="poll"
    key="3"
  >
    <div
      className="pollStatement"
    >
      Ban Poll
    </div>
    <div
      className="pollAnswers"
    >
      <Button
        className="accept"
        color="success"
        id={3}
        onClick={[Function]}
        tag="button"
      >
        Accept
      </Button>
      <Button
        className="decline"
        color="danger"
        id={3}
        onClick={[Function]}
        tag="button"
      >
        Decline
      </Button>
    </div>
  </div>
</div>
`;
